proposals/0000-stream-receipts.md

```diff
@@ -0,0 +1,93 @@
+
+## Overview
+
+In a STREAM connection, a Receipt is generated by the Receiver for every fulfilled ILP packet. Each Receipt on a stream contains a progressively higher amount, representing the total amount received on that stream. So the latest Receipt on a stream replaces any previous Receipts on that stream.
+
+The Verifier SHOULD consider a Receipt valid if:
+
+- The Receipt contents produce an HMAC (using the Receipt Secret as the seed) identical to the Receipt's HMAC field.
+- The amount is greater than any other Receipts with that Receipt Nonce and Stream ID.
+- The Receipt is not stale.
+
+The Verifier SHOULD determine a length of time starting from when a Receipt Nonce is generated that corresponding Receipts will be considered valid. Mechanisms by which the Verifier can check Receipts' staleness include, but are not limited to:
+- persistently storing the Receipt Nonce and a timestamp relative to when the Nonce was generated
+- encoding a timestamp in the Receipt Nonce
+
+If keeping a balance of how much the Sender has paid, the Verifier SHOULD store the Receipt amount under the Receipt Nonce + Stream ID. If there is already an amount stored for that Stream ID, the Verifier SHOULD only credit the Sender the difference in Receipt amounts.
+
+Alteratively, the Verifier MAY forego checking the Receipt amount if they are only concerned with whether or not the Sender paid at all, as opposed to how much.
+
+### Flow
+
+This flow occurs for each STREAM connection with Receipts enabled.
+
+1. Verifier MUST generate a random Receipt Nonce.
+
+2. Verifier MUST communicate the Receipt Nonce and Receipt Secret to the Receiver. The Receipt Secret MUST NOT be communicated to the Sender.
+
+3. Once the STREAM connection is open and the Sender is sending, the Receiver MUST add an additional frame containing a Receipt to each ILP Fulfill packet.
+
+4. Sender MUST give the Receipt directly or indirectly to the Verifier.
+
+5. Verifier MUST verify the Receipt before crediting the Sender with the Receipt amount.
+
+### Stateless Verifier
+
+The Verifier MAY generate the Receipt Secret by producing an HMAC of the Receipt Nonce using a seed (known only to itself) as the HMAC key. This ensures the Receipt Secret will be unique to this STREAM connection and allows the Verifier to reproduce the Receipt Secret from any Receipts containing the Receipt Nonce.
+
+### Stateless Receiver
+
+Before a STREAM connection is established, the Receiver MAY use the Receipt Nonce and Receipt Secret as parameters to generate connection details (ILP Address and shared secret). The Receiver MAY encode the Receipt Nonce and Receipt Secret in the STREAM server ILP Address to allow the STREAM receiver to avoid persisting state. If doing so, the Receipt Secret MUST be encrypted. The Receipt Nonce MUST be unique each time STREAM parameters are generated, since it MUST be unique per STREAM connection.
```

Comment:
`since it MUST be unique per STREAM connection` the receipt nonce must be unique globally